stdlib/socket/src/main/java/org/ballerinalang/stdlib/socket/tcp/SelectorManager.java

```diff
@@ -0,0 +1,238 @@
+import java.nio.channels.SocketChannel;
+import java.util.Iterator;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+
+import static java.nio.channels.SelectionKey.OP_READ;
+
+/**
+ * This will manage the Selector instance and handle the accept, read and write operations.
+ *
+ * @since 0.985.0
+ */
+public class SelectorManager {
+
+    private static final Logger log = LoggerFactory.getLogger(SelectorManager.class);
+
+    private Selector selector;
+    private boolean running = false;
+    private ThreadFactory threadFactory = new BLangThreadFactory("socket-selector");
+    private ExecutorService executor = Executors.newSingleThreadExecutor(threadFactory);
+    private boolean execution = true;
+    private ConcurrentLinkedQueue<ChannelRegisterCallback> registerPendingSockets = new ConcurrentLinkedQueue<>();
+
+    private SelectorManager() throws IOException {
+        selector = Selector.open();
+    }
+
+    /**
+     * This will use to hold the SelectorManager singleton object.
+     */
+    private static class SelectorManagerHolder {
+        private static SelectorManager manager;
+        static {
+            try {
+                manager = new SelectorManager();
+            } catch (IOException e) {
+                throw new SelectorInitializeException("Unable to initialize the selector", e);
+            }
+        }
+    }
+
+    /**
+     * This method will return SelectorManager singleton instance.
+     *
+     * @return {@link SelectorManager} instance
+     * @throws SelectorInitializeException when unable to open a selector
+     */
+    public static SelectorManager getInstance() throws SelectorInitializeException {
+        return SelectorManagerHolder.manager;
+    }
+
+    /**
+     * Add channel to register pending socket queue. Socket registration has to be happen in the same thread
+     * that selector loop execute.
+     *
+     * @param callback A {@link ChannelRegisterCallback} instance which contains the resources,
+     *                      packageInfo and A {@link SelectableChannel}.
+     */
+    public void registerChannel(ChannelRegisterCallback callback) {
+        registerPendingSockets.add(callback);
+        selector.wakeup();
+    }
+
+    /**
+     * Unregister the given client channel from the selector instance.
+     *
+     * @param channel {@link SocketChannel} that about to unregister.
+     */
+    public void unRegisterChannel(SocketChannel channel) {
+        final SelectionKey selectionKey = channel.keyFor(selector);
+        if (selectionKey != null) {
+            selectionKey.cancel();
+        }
+    }
+
+    /**
+     * Start the selector loop.
+     */
+    public void start() {
+        if (running) {
+            return;
+        }
+        executor.execute(this::execute);
+        running = true;
```

Comment:
This method won't be concurrently called right? Otherwise we might need to make last two lines atomic?

Reply:
We can make entire method synchronized. This is not in a critical path and synchronizing method will prevent and abnormal behavior due to concurrency.